### PR TITLE
Added guidance on defining action for critical errors in an endpoint

### PR DESCRIPTION
**Pull request by indualagarsamy**
- Relates to https://github.com/Particular/Requirements/issues/267

- In NServiceBus version 3, if for any reason any of the critical components, such as Timeout persistence, subscription persistence, etc., become unavailable then the endpoint was shutdown.
- Starting from NSB version 4, the core will shutdown the bus, but not the process. It is the responsibility of the host to listen to critical errors and take appropriate action.
- For our self hosting customers, our guidance did not do a good job of letting them know that they need to do when such things occurred. 

This PR adds guidance on how to define an action so that the user can take better action when such failures occur.

@Particular/guidance-support please review

### CODE REVIEW

File: nservicebus/hosting/handling-failures.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+---
+title: Handling failures 
+summary: How to subscribe to notifications regarding critical errors which adversely affect messaging in your endpoint.
+tags:
+- NServiceBus Host
+- Self Hosting
+---
+
+NServiceBus offers an error event handler that you can use in order to receive notifications when something undesirable happens in the endpoint which involves the ability to receive or send messages. E.g. When your persistence storage that is essential for messaging goes down.
```
Comment (SeanFeldman) [inline nservicebus/hosting/handling-failures.md]: >something undesirable happens

should it be "critical error takes place" to signify importance of this notifications?

File: nservicebus/hosting/handling-failures.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+<!-- import CustomHostErrorHandlingSubscription -->
+
+Next you define what action you want to take when this scenario occurs:
+
+<!-- import CustomHostErrorHandlingAction -->
+
+## Why should you do this?
+
+- If you are using NServiceBus Host, and you wish to take a custom action before the endpoint process is killed.
+- If you are self hosting, by defining this action, you can get the same behavior as that of NServiceBus.host by calling `Environment.FailFast` which will kill the process. If this is a website and you don't wish to terminate the process, you can create a Fatal log entry and then dispose the bus, which you can then monitor and restart the website accordingly when convenient. By defining this action, if you're using [ServicePulse]() to monitor your web endpoint, then by doing so, it would appear `InActive`.
```
Comment (SeanFeldman) [inline nservicebus/hosting/handling-failures.md]: Link to SP is missing

File: nservicebus/hosting/handling-failures.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+If you choose to not kill the process and just dispose the bus, please be aware that any `bus.Send` operations will result in `ObjectDisposedException` and you will need to cater for it.
+
+NOTE:  When self hosting using NServiceBus Version4 and you did not explicitly define this action, while the bus was shutdown it wasn't disposed. Therefore the endpoint allowed you to send messages. However since the bus was shutdown, it did not process any messages that arrived in the queue. Therefore the process was running but not processing messages. By defining this action, you can now be aware of this situation and dispose the bus and optionally terminate the process based on your needs.
```
Comment (SeanFeldman) [inline nservicebus/hosting/handling-failures.md]: version 4 (space)

**Comment by SimonCropp** [comment]
@indualagarsamy @SeanFeldman i re worked this a little. can you both review please

**Comment by SeanFeldman** [comment]
`bus.Dispose()` in critical error handler demonstration is a good idea.

**Pull request merged**